### PR TITLE
KEP-127: Add UserNamespacesPodSecurityStandards e2e test

### PR DESCRIPTION
**Pull request by saschagrunert**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:
Adding a e2e test for the functionality added in https://github.com/kubernetes/kubernetes/pull/118760.

#### Which issue(s) this PR fixes:

None

#### Special notes for your reviewer:
cc @mrunalp @rata @giuseppe 
#### Does this PR introduce a user-facing change?

```release-note
None
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
None
```

### CODE REVIEW

File: test/e2e_node/user_namespaces_test.go [hunk 1]
```diff
@@ -0,0 +1,97 @@
+//go:build linux
```
Comment (rata) [inline test/e2e_node/user_namespaces_test.go:1]: Just curious, why here? We have other e2e tests here: https://github.com/kubernetes/kubernetes/commit/b1eaf6a2d978d460d5d8f274102851925afff53c

I'm completely okay if this test makes sense here, just curious about the reason

File: test/e2e_node/user_namespaces_test.go [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+					HostUsers:       ptr.To(false),
+					SecurityContext: &v1.PodSecurityContext{},
+					Containers: []v1.Container{
+						{
+							Name:    name,
+							Image:   busyboxImage,
+							Command: []string{"echo", "test"},
+							SecurityContext: &v1.SecurityContext{
+								AllowPrivilegeEscalation: ptr.To(false),
+								Capabilities:             &v1.Capabilities{Drop: []v1.Capability{"ALL"}},
```
Comment (rata) [inline test/e2e_node/user_namespaces_test.go:65]: Why is this needed or why do we want this?

File: test/e2e_node/user_namespaces_test.go [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+			pod := newPod()
+
+			user := ptr.To(int64(1000))
+			pod.Spec.SecurityContext.RunAsUser = user
+			pod.Spec.Containers[0].SecurityContext.RunAsUser = user
+
+			e2eoutput.TestContainerOutput(ctx, f, "RunAsUser", pod, 0, []string{"test"})
+		})
+
+		ginkgo.It("should allow pod with RunAsNonRoot", func(ctx context.Context) {
```
Comment (rata) [inline test/e2e_node/user_namespaces_test.go:84]: These two tests seem redundant, both are testing it can run as UID 1000. The second one is setting the runAsNonRoot to true, but the base pod we are using already has that...

Am I missing something?

**Review by rata — COMMENTED** [review]
I think the tests are not testing the functionality we are adding. Don't we want to test that you **can** run as root, with the PSS restricted and the feature gate enabled?

Am I missing something?

File: test/e2e_node/user_namespaces_test.go (hunk 1 above)
Comment (saschagrunert) [reply test/e2e_node/user_namespaces_test.go:1]: My goal was to enable the UserNamespacesSupport and UserNamespacesPodSecurityStandards features in the tests, but that does not seem to work with the API server.

File: test/e2e_node/user_namespaces_test.go (hunk 3 above)
Comment (saschagrunert) [reply test/e2e_node/user_namespaces_test.go]: Moved it into a single test.

File: test/e2e_node/user_namespaces_test.go (hunk 2 above)
Comment (saschagrunert) [reply test/e2e_node/user_namespaces_test.go]: The capabilities are required for the restricted pod security.

File: test/e2e/feature/feature.go [hunk 4]
```diff
@@ -120,6 +120,7 @@ var (
 	UDP                                     = framework.WithFeature(framework.ValidFeatures.Add("UDP"))
 	Upgrade                                 = framework.WithFeature(framework.ValidFeatures.Add("Upgrade"))
 	UserNamespacesSupport                   = framework.WithFeature(framework.ValidFeatures.Add("UserNamespacesSupport"))
+	UserNamespacesPodSecurityStandards      = framework.WithFeature(framework.ValidFeatures.Add("UserNamespacesPodSecurityStandards"))
```
Comment (rata) [inline test/e2e/feature/feature.go:123]: Don't we need user namespaces support too?

File: test/e2e_node/user_namespaces_test.go (hunk 2 above)
Comment (rata) [reply test/e2e_node/user_namespaces_test.go]: Oh, we require no capability at all? Not even the defaults? Is that what is otherwise required by the PSS for the non userns case?

Comment (saschagrunert) [reply test/e2e_node/user_namespaces_test.go]: If I remove that from the test, the admission states:

```
Error creating Pod: pods "pod-user-namespaces-pss-df7bfa73-af66-48a8-b408-b80ce5104401" is forbidden: violates PodSecurity "restricted:latest": unrestricted capabilities (container "pod-user-namespaces-pss-df7bfa73-af66-48a8-b408-b80ce5104401" must set securityContext.capabilities.drop=["ALL"])
```

File: test/e2e/feature/feature.go (hunk 4 above)
Comment (saschagrunert) [reply test/e2e/feature/feature.go:123]: I have `[Feature:UserNamespacesSupport] [Feature:UserNamespacesPodSecurityStandards]` in the tests, is that sufficient?

Comment (rata) [reply test/e2e/feature/feature.go:123]: I think it might be, but I'm not sure tbh

**Review by rata — APPROVED** [review]
LGTM, thanks!

File: test/e2e_node/user_namespaces_test.go (hunk 2 above)
Comment (rata) [reply test/e2e_node/user_namespaces_test.go]: Perfect then, thanks!

File: test/e2e/common/node/security_context.go [hunk 5]
```diff
@@ -633,6 +634,39 @@ var _ = SIGDescribe("Security Context", func() {
 	})
 })
 
+var _ = SIGDescribe("User Namespaces for Pod Security Standards [Feature:UserNamespacesSupport] [Feature:UserNamespacesPodSecurityStandards] [LinuxOnly]", func() {
```
Comment (SergeyKanzhelev) [inline test/e2e/common/node/security_context.go:637]: should you use the newly added valid feature variable here (`UserNamespacesPodSecurityStandards`)?

Comment (saschagrunert) [reply test/e2e/common/node/security_context.go:637]: Only `[Feature:UserNamespacesPodSecurityStandards]`? I was assuming that we use `[Feature:UserNamespacesSupport]` as well because other tests are referring to it, too.

**Review by SergeyKanzhelev — APPROVED** [review]
/lgtm
/approve

**Review by SergeyKanzhelev — APPROVED** [review]
/lgtm
/approve

**Comment by saschagrunert** [comment]
Had to rebase on top of the latest `master` branch.

**Comment by pohly** [comment]
@saschagrunert: please fix these errors:
```
k8s.io/kubernetes/test/e2e/common/node/security_context.go:639: [Feature:UserNamespacesPodSecurityStandards] in plain text is deprecated and must be added through WithFeature(UserNamespacesPodSecurityStandards) instead
k8s.io/kubernetes/test/e2e/common/node/security_context.go:639: [Feature:UserNamespacesSupport] in plain text is deprecated and must be added through WithFeature(UserNamespacesSupport) instead 
 ```
 
 You can check locally with `hack/verify-e2e-suites.sh`.

File: test/e2e/feature/feature.go [hunk 6]
```diff
@@ -328,6 +328,9 @@ var (
 	// TODO: document the feature (owning SIG, when to use this feature for a test)
 	UserNamespacesSupport = framework.WithFeature(framework.ValidFeatures.Add("UserNamespacesSupport"))
 
+	// TODO: document the feature (owning SIG, when to use this feature for a test)
```
Comment (pohly) [inline test/e2e/feature/feature.go:331]: Please add documentation together the new feature.

@SergeyKanzhelev: would you have pointed out the same thing if I hadn't gotten involved?

I'm not sure how to ensure that developers do this without being reminded. Suggestions welcome :thinking:

**Comment by saschagrunert** [comment]
> @saschagrunert: please fix these errors:
> 
> ```
> k8s.io/kubernetes/test/e2e/common/node/security_context.go:639: [Feature:UserNamespacesPodSecurityStandards] in plain text is deprecated and must be added through WithFeature(UserNamespacesPodSecurityStandards) instead
> k8s.io/kubernetes/test/e2e/common/node/security_context.go:639: [Feature:UserNamespacesSupport] in plain text is deprecated and must be added through WithFeature(UserNamespacesSupport) instead 
> ```
> 
> You can check locally with `hack/verify-e2e-suites.sh`.

Thank you for the hint, fixed as suggested

File: test/e2e/feature/feature.go (hunk 6 above)
Comment (SergeyKanzhelev) [reply test/e2e/feature/feature.go]: > @SergeyKanzhelev: would you have pointed out the same thing if I hadn't gotten involved?

Yes, after the PR with `TODO:`s was merged.

**Review by SergeyKanzhelev — APPROVED** [review]
/lgtm
/approve
